data/data/bootstrap/files/usr/local/bin/bootkube.sh.template

```diff
@@ -96,6 +96,84 @@ then
 	touch config-bootstrap.done
 fi
 
+# Rendering MCO first so machines can be boot before we wait for their
+# mDNS information
+if [ ! -f mco-bootstrap.done ]
+then
+	echo "Rendering MCO manifests..."
+
+	rm -rf mco-bootstrap
+
+	# shellcheck disable=SC2154
+	podman run \
+		--quiet \
+		--user 0 \
+		--volume "$PWD:/assets:z" \
+		"${MACHINE_CONFIG_OPERATOR_IMAGE}" \
+		bootstrap \
+			--etcd-ca=/assets/tls/etcd-ca-bundle.crt \
+			--etcd-metric-ca=/assets/tls/etcd-metric-ca-bundle.crt \
+			--root-ca=/assets/tls/root-ca.crt \
+			--kube-ca=/assets/tls/kube-apiserver-complete-client-ca-bundle.crt \
+			--config-file=/assets/manifests/cluster-config.yaml \
+			--dest-dir=/assets/mco-bootstrap \
+			--pull-secret=/assets/manifests/openshift-config-secret-pull-secret.yaml \
+			--etcd-image=${MACHINE_CONFIG_ETCD_IMAGE} \
+			--setup-etcd-env-image=${MACHINE_CONFIG_SETUP_ETCD_ENV_IMAGE} \
+			--kube-client-agent-image=${MACHINE_CONFIG_KUBE_CLIENT_AGENT_IMAGE} \
+			--machine-config-controller-image=${MACHINE_CONFIG_CONTROLLER_IMAGE} \
+			--machine-config-server-image=${MACHINE_CONFIG_SERVER_IMAGE} \
+			--machine-config-daemon-image=${MACHINE_CONFIG_DAEMON_IMAGE} \
+                        --machine-config-operator-image=${MACHINE_CONFIG_OPERATOR_IMAGE} \
+			--machine-config-oscontent-image=${MACHINE_CONFIG_OSCONTENT} \
+			--infra-image=${MACHINE_CONFIG_INFRA_IMAGE} \
+			--cloud-config-file=/assets/manifests/cloud-provider-config.yaml
+
+	# Bootstrap MachineConfigController uses /etc/mcc/bootstrap/manifests/ dir to
+	# 1. read the controller config rendered by MachineConfigOperator
+	# 2. read the default MachineConfigPools rendered by MachineConfigOperator
+	# 3. read any additional MachineConfigs that are needed for the default MachineConfigPools.
+	mkdir --parents /etc/mcc/bootstrap /etc/mcs/bootstrap /etc/kubernetes/manifests
+	cp mco-bootstrap/bootstrap/manifests/* /etc/mcc/bootstrap/
+	cp openshift/* /etc/mcc/bootstrap/
+	cp auth/kubeconfig-kubelet /etc/mcs/kubeconfig
+	cp mco-bootstrap/bootstrap/machineconfigoperator-bootstrap-pod.yaml /etc/kubernetes/manifests/
+	cp mco-bootstrap/manifests/* manifests/
+
+	# /etc/ssl/mcs/tls.{crt, key} are locations for MachineConfigServer's tls assets.
+	mkdir --parents /etc/ssl/mcs/
+	cp tls/machine-config-server.crt /etc/ssl/mcs/tls.crt
+	cp tls/machine-config-server.key /etc/ssl/mcs/tls.key
+
+	touch mco-bootstrap.done
+fi
+
+echo "Waiting for etcd member names to be available from SRV ..."
```

Comment:
Why do we care?

Reply:
@celebdor @cybertron

Reply:
We needed this to get the etcd members. If they aren't available from DNS yet we can't continue.

Reply:
How is that it is working fine for now, and bare metal needs this?

Reply:
>yet we can't continue.

continue to what?

is there lack of resiliency in your scripts, if yes please fix them.

Reply:
As I understand it, the other platforms are not actually looking up the etcd member names from DNS. They're assuming etc-0,1,2. We can't do that because our hostnames are set by site DHCP, which we don't control, and it's problematic to map arbitrary hostnames to indices in a stable way so that they won't change if, for example, a master goes down and that DNS entry disappears.

We can't continue without the DNS entries because the next step in bootstrapping requires us to pass the etcd server urls, which we don't have until they're available from DNS. It's simply a dependency, not a lack of resiliency.

Reply:
> As I understand it, the other platforms are not actually looking up the etcd member names from DNS. They're assuming etc-0,1,2.

that's not true,
installer creates those dns records (etcd-$idx) and points them to the hostnames, and passes that (etcd-$idx) to the operators.

Reply:
It sounds like the main point is that we need to be passing these names from the installer, not finding them at bootstrap time. That seems doable, so we'll look into it.

Reply:
@celebdor Can you describe the alternatives you have in mind?  On the surface, the contents of the SRV record look pretty simple and could be passed in via the installer instead of dynamically discovered:

```
[core@master-0 ~]$ host -t SRV "_etcd-server-ssl._tcp.ostest.test.metalkube.org"
_etcd-server-ssl._tcp.ostest.test.metalkube.org has SRV record 0 10 2380 etcd-2.ostest.test.metalkube.org.
_etcd-server-ssl._tcp.ostest.test.metalkube.org has SRV record 0 10 2380 etcd-1.ostest.test.metalkube.org.
_etcd-server-ssl._tcp.ostest.test.metalkube.org has SRV record 0 10 2380 etcd-0.ostest.test.metalkube.org.
```

These hostnames are not the ones that are provided by the external DHCP server, so assumptions about the DHCP provided hostname don't seem to be relevant here.

O thing that is different with this platform is that the related A records are not created in advance.  For other platforms, they are created in advance before any of this runs.  In the case of the baremetal platform, all of this internal DNS resolution is based on discoery using mDNS, so these `etcd-$N` records will not be resolvable until the masters come up and run their mDNS publisher.

Is the real issue here instead that something needs to not run until these records are resolvable?

Reply:
I guess since https://github.com/openshift-metal3/dev-scripts/commit/e2b12a7386af18dd189d46f09412489fd1601509 it would be possible to hard-code these entries. As long as we're enforcing specific hostnames we should know ahead of time what the etcd entries will look like.

Ideally we _wouldn't_ enforce hostnames this way and would dynamically look them up, but if we're willing to push that additional requirement onto users we can simplify this a lot.

Reply:
@cybertron pointed me to https://github.com/openshift-metal3/dev-scripts/blob/master/assets/files/etc/kubernetes/manifests/mdns-publisher.yaml#L62 which explains how the etcd hostnames are derived from the DHCP provided hostname.  If we statically define this as `etcd-$N`, then we're imposing a naming pattern restriction on master hostnames.

Obviously a hostname pattern restriction isn't ideal, but that seems like a reasonable next step here, and then a todo item to remove that as a restriction for baremetal IPI in the future.

That will at least allow us to remove this magic from bootkube.

Reply:
Removed in openshift-metal3/kni-installer#130 so it'll be gone when this PR is refreshed